File: hermes-kotlin/src/main/kotlin/hermes/Embedding.kt
package hermes

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

private const val DEFAULT_MODEL = "text-embedding-004"
private const val DEFAULT_DIMENSION = 768
private const val DEFAULT_RPM = 60

class EmbeddingGenerator(
    private val apiKey: String = System.getenv("GEMINI_API_KEY")
        ?: throw IllegalStateException("GEMINI_API_KEY environment variable not set"),
    private val model: String = System.getenv("GEMINI_EMBEDDING_MODEL") ?: DEFAULT_MODEL,
    private val client: OkHttpClient = OkHttpClient()
) {
    companion object {
        fun dimension(): Int = DEFAULT_DIMENSION
    }

    @Serializable
    private data class EmbeddingRequest(
        val model: String,
        val content: EmbeddingContent
    )

    @Serializable
    private data class EmbeddingContent(val parts: List<EmbeddingPart>)

    @Serializable
    private data class EmbeddingPart(val text: String)

    @Serializable
    private data class EmbeddingResponse(val embedding: EmbeddingValues)

    @Serializable
    private data class EmbeddingValues(val values: List<Float>)

    private val json = Json { ignoreUnknownKeys = true }

    fun generateEmbedding(text: String): List<Float> {
        val url = "https://generativelanguage.googleapis.com/v1beta/models/$model:embedContent?key=$apiKey"

        val requestBody = json.encodeToString(
            EmbeddingRequest.serializer(),
            EmbeddingRequest(
                model = "models/$model",
                content = EmbeddingContent(parts = listOf(EmbeddingPart(text = text)))
            )
        )

        val request = Request.Builder()
            .url(url)
            .post(requestBody.toRequestBody("application/json".toMediaType()))
            .build()

        val response = client.newCall(request).execute()
        val body = response.body?.string() ?: throw RuntimeException("Empty response from embedding API")

        if (!response.isSuccessful) {
            throw RuntimeException("Embedding API returned ${response.code}: $body")
        }

        val parsed = json.decodeFromString(EmbeddingResponse.serializer(), body)
        return parsed.embedding.values
    }

    fun generateEmbeddings(texts: List<String>): List<List<Float>> {
        return texts.map { generateEmbedding(it) }
    }
}

File: src/embedding.rs
use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::env;
use std::sync::Arc;
use tokio::sync::Semaphore;

const DEFAULT_MODEL: &str = "text-embedding-004";
const DEFAULT_DIMENSION: usize = 768;
const DEFAULT_RPM: usize = 60;

#[derive(Clone)]
pub struct EmbeddingGenerator {
    api_key: String,
    model: String,
    client: reqwest::Client,
    rate_limiter: Arc<Semaphore>,
}

#[derive(Debug, Serialize)]
struct EmbeddingRequest {
    model: String,
    content: EmbeddingContent,
}

#[derive(Debug, Serialize)]
struct EmbeddingContent {
    parts: Vec<EmbeddingPart>,
}

#[derive(Debug, Serialize)]
struct EmbeddingPart {
    text: String,
}

#[derive(Debug, Deserialize)]
struct EmbeddingResponse {
    embedding: EmbeddingValues,
}

#[derive(Debug, Deserialize)]
struct EmbeddingValues {
    values: Vec<f32>,
}

impl EmbeddingGenerator {
    pub fn new() -> Result<Self> {
        let api_key = env::var("GEMINI_API_KEY")
            .context("GEMINI_API_KEY environment variable not set")?;
        let model = env::var("GEMINI_EMBEDDING_MODEL")
            .unwrap_or_else(|_| DEFAULT_MODEL.to_string());
        let rpm: usize = env::var("EMBEDDING_RPM")
            .ok()
            .and_then(|v| v.parse().ok())
            .unwrap_or(DEFAULT_RPM);

        Ok(Self {
            api_key,
            model,
            client: reqwest::Client::new(),
            rate_limiter: Arc::new(Semaphore::new(rpm)),
        })
    }

    pub fn dimension() -> usize {
        DEFAULT_DIMENSION
    }

    pub async fn generate_embedding(&self, text: &str) -> Result<Vec<f32>> {
        let _permit = self.rate_limiter.acquire().await
            .map_err(|e| anyhow::anyhow!("Rate limiter closed: {e}"))?;

        let url = format!(
            "https://generativelanguage.googleapis.com/v1beta/models/{}:embedContent?key={}",
            self.model, self.api_key
        );

        let request = EmbeddingRequest {
            model: format!("models/{}", self.model),
            content: EmbeddingContent {
                parts: vec![EmbeddingPart {
                    text: text.to_string(),
                }],
            },
        };

        let response = self
            .client
            .post(&url)
            .json(&request)
            .send()
            .await
            .context("Failed to call embedding API")?;

        let status = response.status();
        if !status.is_success() {
            let body = response.text().await.unwrap_or_default();
            anyhow::bail!("Embedding API returned {status}: {body}");
        }

        let parsed: EmbeddingResponse = response
            .json()
            .await
            .context("Failed to parse embedding response")?;

        Ok(parsed.embedding.values)
    }

    pub async fn generate_embeddings(&self, texts: &[&str]) -> Result<Vec<Vec<f32>>> {
        let mut results = Vec::with_capacity(texts.len());
        for text in texts {
            let embedding = self.generate_embedding(text).await?;
            results.push(embedding);
        }
        Ok(results)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn dimension_is_768() {
        assert_eq!(EmbeddingGenerator::dimension(), 768);
    }
}
